### PR TITLE
Add support for gRPC streams.

### PR DESCRIPTION
**Pull request by Shikugawa**
Add gRPC stream support for Rust SDK. It includes any workaround around callout token table caused by #100 is not landed to master.

### CODE REVIEW

File: src/traits.rs [hunk 1]
```diff
@@ -119,6 +119,45 @@ pub trait Context {
...
+
+    fn on_grpc_stream_receive_initial_metadata(&mut self, _token_id: u32, _headers: u32) {}
+
+    fn on_grpc_stream_receive_trailing_metadata(&mut self, _token_id: u32, _trailers: u32) {}
+
+    fn on_grpc_stream_receive_body(&mut self, _token_id: u32, _response_size: usize) {}
+
+    fn on_grpc_stream_close(&mut self, _token_id: u32, _status_code: u32) {}
+
+    fn get_grpc_call_initial_metadata(&self) -> Vec<(String, Vec<u8>)> {
```
Comment (PiotrSikora) [inline src/traits.rs:153]: Nit: `get_grpc_stream_initial_metadata`

File: src/traits.rs [hunk 2]
```diff
@@ -119,6 +119,45 @@ pub trait Context {
...
+        hostcalls::get_map_bytes(MapType::GrpcReceiveInitialMetadata).unwrap()
+    }
+
+    fn get_grpc_call_trailing_metadata(&self) -> Vec<(String, Vec<u8>)> {
```
Comment (PiotrSikora) [inline src/traits.rs:157]: Nit: `get_grpc_stream_trailing_metadata`

File: src/traits.rs [hunk 3]
```diff
@@ -119,6 +119,45 @@ pub trait Context {
...
+
+    fn grpc_stream_close(&self, token_id: u32) -> Result<(), Status> {
+        hostcalls::grpc_stream_close(token_id)
+    }
+
+    fn on_grpc_stream_receive_initial_metadata(&mut self, _token_id: u32, _headers: u32) {}
+
+    fn on_grpc_stream_receive_trailing_metadata(&mut self, _token_id: u32, _trailers: u32) {}
+
+    fn on_grpc_stream_receive_body(&mut self, _token_id: u32, _response_size: usize) {}
```
Comment (PiotrSikora) [inline src/traits.rs:149]: Nit: `on_grpc_stream_message` and `_message_size`.

File: src/traits.rs [hunk 4]
```diff
@@ -119,6 +119,45 @@ pub trait Context {
...
+        end_stream: bool,
+    ) -> Result<(), Status> {
+        hostcalls::grpc_stream_send(token_id, message, end_stream)
+    }
+
+    fn grpc_stream_close(&self, token_id: u32) -> Result<(), Status> {
+        hostcalls::grpc_stream_close(token_id)
+    }
+
+    fn on_grpc_stream_receive_initial_metadata(&mut self, _token_id: u32, _headers: u32) {}
```
Comment (PiotrSikora) [inline src/traits.rs:145]: Nit: drop `receive`, it's doesn't add any value here, so `on_grpc_stream_initial_metadata`.

Also, replace `_headers` with `_num_elements`

File: src/traits.rs [hunk 5]
```diff
@@ -119,6 +119,45 @@ pub trait Context {
...
+
+    fn on_grpc_stream_receive_trailing_metadata(&mut self, _token_id: u32, _trailers: u32) {}
```
Comment (PiotrSikora) [inline src/traits.rs:147]: Nit: drop `receive`, it's doesn't add any value here, so `on_grpc_stream_trailing_metadata`.

Also, replace `_trailers` with `_num_elements`

File: src/traits.rs [hunk 6]
```diff
@@ -119,6 +119,45 @@ pub trait Context {
...
+    fn on_grpc_stream_close(&mut self, _token_id: u32, _status_code: u32) {}
+
+    fn get_grpc_call_initial_metadata(&self) -> Vec<(String, Vec<u8>)> {
+        hostcalls::get_map_bytes(MapType::GrpcReceiveInitialMetadata).unwrap()
+    }
+
+    fn get_grpc_call_trailing_metadata(&self) -> Vec<(String, Vec<u8>)> {
+        hostcalls::get_map_bytes(MapType::GrpcReceiveTrailingMetadata).unwrap()
+    }
+
```
Comment (PiotrSikora) [inline src/traits.rs:160]: Could you please re-order them in order of use, i.e.:
- `open_grpc_stream`
- `on_grpc_stream_initial_metadata`
- `get_grpc_stream_initial_metadata`
- `send_grpc_stream_message`
- `on_grpc_stream_message`
- `get_grpc_stream_message` (currently missing)
- `on_grpc_stream_trailing_metadata`
- `get_grpc_stream_trailing_metadata`
- `close_grpc_stream`
- `on_grpc_stream_close`

**Review by PiotrSikora — COMMENTED** [review]
Mostly a style-pass, otherwise looks fine (sans missing `get_grpc_stream_message`).

Also, please mark this PR as ready for review when it's ready.

File: src/dispatcher.rs [hunk 7]
```diff
@@ -400,46 +417,117 @@ impl Dispatcher {
...
-            stream.on_grpc_call_response(token_id, status_code, 0);
+            stream.on_grpc_stream_receive_trailing_metadata(token_id, trailers);
         } else if let Some(root) = self.roots.borrow_mut().get_mut(&context_id) {
             self.active_id.set(context_id);
             hostcalls::set_effective_context(context_id).unwrap();
-            root.on_grpc_call_response(token_id, status_code, 0);
+            root.on_grpc_stream_receive_trailing_metadata(token_id, trailers);
         }
     }
 }
```
[the remark this answers is not in the document]
Comment (Shikugawa) [reply src/dispatcher.rs:533]: done

File: src/traits.rs (hunk 6 above)
Comment (Shikugawa) [reply src/traits.rs:162]: done

File: src/hostcalls.rs [hunk 8]
```diff
@@ -893,4 +981,25 @@ mod utils {
         }
         map
     }
+
+    pub(super) fn deserialize_bytes_map(bytes: &[u8]) -> Vec<(String, Vec<u8>)> {
```
[the remark this answers is not in the document]
Comment (Shikugawa) [reply src/hostcalls.rs:1003]: changed

File: src/dispatcher.rs [hunk 9]
```diff
@@ -571,11 +657,25 @@ pub extern "C" fn proxy_on_http_call_response(
...
+    DISPATCHER.with(|dispatcher| dispatcher.on_grpc_receive_initial_metadata(token_id, headers))
+}
+
 #[no_mangle]
 pub extern "C" fn proxy_on_grpc_receive(_context_id: u32, token_id: u32, response_size: usize) {
     DISPATCHER.with(|dispatcher| dispatcher.on_grpc_receive(token_id, response_size))
 }
 
+#[no_mangle]
+pub extern "C" fn proxy_on_grpc_trailing_metadata(_context_id: u32, token_id: u32, trailers: u32) {
```
Comment (PiotrSikora) [inline src/dispatcher.rs:675]: This must be called `proxy_on_grpc_receive_trailing_metadata` (missing `receive`).

File: src/dispatcher.rs [hunk 10]
```diff
@@ -399,47 +416,116 @@ impl Dispatcher {
...
             .borrow_mut()
-            .remove(&token_id)
+            .get(&token_id)
             .expect("invalid token_id");
 
         if let Some(http_stream) = self.http_streams.borrow_mut().get_mut(&context_id) {
             self.active_id.set(context_id);
             hostcalls::set_effective_context(context_id).unwrap();
-            http_stream.on_grpc_call_response(token_id, status_code, 0);
+            http_stream.on_grpc_stream_initial_metadata(token_id, trailers);
```
Comment (PiotrSikora) [inline src/dispatcher.rs:486]: `on_grpc_stream_trailing_metadata` (`s/initial/trailing/`).

File: src/dispatcher.rs [hunk 11]
```diff
@@ -399,47 +416,116 @@ impl Dispatcher {
...
         } else if let Some(stream) = self.streams.borrow_mut().get_mut(&context_id) {
             self.active_id.set(context_id);
             hostcalls::set_effective_context(context_id).unwrap();
-            stream.on_grpc_call_response(token_id, status_code, 0);
+            stream.on_grpc_stream_initial_metadata(token_id, trailers);
```
Comment (PiotrSikora) [inline src/dispatcher.rs:490]: `on_grpc_stream_trailing_metadata` (`s/initial/trailing/`).

File: src/dispatcher.rs [hunk 12]
```diff
@@ -399,47 +416,116 @@ impl Dispatcher {
...
         } else if let Some(root) = self.roots.borrow_mut().get_mut(&context_id) {
             self.active_id.set(context_id);
             hostcalls::set_effective_context(context_id).unwrap();
-            root.on_grpc_call_response(token_id, status_code, 0);
+            root.on_grpc_stream_initial_metadata(token_id, trailers);
```
Comment (PiotrSikora) [inline src/dispatcher.rs:494]: `on_grpc_stream_trailing_metadata` (`s/initial/trailing/`).

File: src/dispatcher.rs [hunk 13]
```diff
@@ -571,11 +657,25 @@ pub extern "C" fn proxy_on_http_call_response(
...
+    DISPATCHER.with(|dispatcher| dispatcher.on_grpc_receive_initial_metadata(token_id, headers))
+}
+
 #[no_mangle]
 pub extern "C" fn proxy_on_grpc_receive(_context_id: u32, token_id: u32, response_size: usize) {
     DISPATCHER.with(|dispatcher| dispatcher.on_grpc_receive(token_id, response_size))
 }
 
+#[no_mangle]
+pub extern "C" fn proxy_receive_on_grpc_trailing_metadata(_context_id: u32, token_id: u32, trailers: u32) {
```
Comment (PiotrSikora) [inline src/dispatcher.rs:675]: This is still wrong, it must be `proxy_on_grpc_receive_trailing_metadata` (`receive` in the wrong place).

File: src/dispatcher.rs [hunk 14]
```diff
@@ -571,11 +657,29 @@ pub extern "C" fn proxy_on_http_call_response(
...
+    DISPATCHER.with(|dispatcher| dispatcher.on_grpc_receive_initial_metadata(token_id, headers))
+}
+
 #[no_mangle]
 pub extern "C" fn proxy_on_grpc_receive(_context_id: u32, token_id: u32, response_size: usize) {
     DISPATCHER.with(|dispatcher| dispatcher.on_grpc_receive(token_id, response_size))
 }
 
+#[no_mangle]
+pub extern "C" fn proxy_on_receive_grpc_trailing_metadata(
```
Comment (PiotrSikora) [inline src/dispatcher.rs:675]: This is still worng... It _has_ to be `proxy_on_grpc_receive_trailing_metadata` to match the signature from the host.

Please copy & paste this function name.

**Review by PiotrSikora — CHANGES_REQUESTED** [review]

File: src/dispatcher.rs (hunk 14 above)
Comment (Shikugawa) [reply src/dispatcher.rs]: oh. sorry...

**Review by PiotrSikora — APPROVED** [review]
Thank you!

**Pull request merged**